### ISSUE TITLE
Kustomize labels configuration options

### ISSUE THREAD

**Issue by pokgak** [argoproj/argo-cd#15283]
# Summary

Expose the kustomize `labels` [configuration options](https://argo-cd.readthedocs.io/en/stable/user-guide/kustomize/) in kustomization.yaml to allow adding labels to resources but skips adding to the matchLabels and selector fields in resources. The commonLabels options automatically adds the labels to both matchLabels and selector fields and there is no way to turn it off. This prevents using a dynamic value for the label as the matchLabels field is immutable and cannot be changed once created.

The label configuration options is better since it allow users to configure whether or not to add the label to the selector and template.

# Motivation

I'm currently reading the image tag through ApplicationSet templating using git files generator. The image tag is read from a file and passed to the application using kustomize `images` configuration options. Using the same image tag value, I also want to add a label to my Deployment resource.

Using the `commonLabels` configuration option, it also adds the label to the matchSelector. Since my image tag value always changes, it will produce error when syncing since the matchLabel field is actually immutable and cannot be changed once created.

The `labels` configuration option is the better way to achive what I'm trying to do but it is not exposed by ArgoCD Application resource as part of the kustomize configuration option currently.

# Proposal

Add new `labels` configuration option to ArgoCD Application kustomize fields.

Documentation on kustomize `labels`: https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/labels/

**Comment by mtrin** [comment]
Also having the same problem. In fact, is there any reason not to have every kustomize top level field available by default?
https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/

**Comment by dmavis** [comment]
To echo the earlier comments on this issue, the additional of the `labels` configuration option to Argo CD Application kustomize fields would be very useful to our platform team. Plus, `commonLabels` have been deprecated: https://github.com/kubernetes-sigs/kustomize/pull/5464

**Comment by joebowbeer** [comment]
To be clear, this can be achieved now using new `patches` or `components`?

**Comment by dmavis** [comment]
> To be clear, this can be achieved now using new `patches` or `components`?

Indeed, I had missed that the `patches` Kustomize option was recently added to the Argo CD Application in v2.9.0. However, it would be more convenient and intuitive if the `labels` Kustomize option was also added. Thanks for the heads up on using `patches` for now.

**Comment by mtrin** [comment]
Good to know patches is available but it can make things more complex as you'd need a patch for anything that has selectors.
`labels` feature in Kustomize already figures that out

**Comment by joebowbeer** [comment]
Please also support `replacements`

It [behaves more intuitively](https://github.com/kubernetes-sigs/kustomize/issues/4099) when it is top-level and not embedded in a component

**Comment by dantman** [comment]
I had this in my previous kustomize config, processed by envsubst instead of Argo CD, to add an environment label that's not included in selectors.
```
namespace: $ENVIRONMENT
labels:
  - pairs:
      environment: $ENVIRONMENT
    includeTemplates: true
```

When I switched to an Argo CD application created by Terraform. It was rather disappointing that I gained the ability to set the `namespace` directly from Terraform variables but could not do the same for adding labels that were not included in selectors.

**Comment by gnarroway-bullish** [comment]
Haven't tested it myself, but can you directly use the transformers instead of the top level option? 

e.g. 
- labels: https://github.com/keleustes/kustomize/tree/allinone/examples/issues/issue_0817_b
- replacements: https://github.com/kubernetes-sigs/kustomize/issues/4476#issuecomment-1041904099

**Comment by crenshaw-dev** [comment]
Proposal:

```yaml
spec:
  source:
    kustomize:
      commonLabels:
        a: b
      labelsWithoutSelector: true
      # Proposed new field:
      includeTemplates: true
```

The new field would correspond to the CLI's `--include-templates` flag. Under the hood, when you use `--without-selector` and `--include-templates`, Kustomize populates a `labels` config.

It's unfortunate that our current CRD structure mirrors the CLI flags rather than the kustomization.yaml schema, but it's probably best to stick with that until we cut a new CRD apiVersion.

Kustomize will throw an error if `includeTemplates` is `true` but `labelsWithoutSelector` is `false`. We could try to be smart and infer `labelsWithoutSelector=true` from `includeTemplates=true`. But I lean slightly toward matching the CLI behavior 1:1 and just requiring the user to always set both.
